README.MD
# Tweet Classification for Personal Health Mention
This application classifies user input tweets into their relevant health event groups. The [ERNIE 2.0 model](https://huggingface.co/dibsondivya/ernie-phmtweets-sutd) used for classification has been trained on data found from an [Emory University Study on Detection of Personal Health Mentions in Social Media paper](https://arxiv.org/pdf/1802.09130v2.pdf), and fine-tuned to perform with 88.5% accuracy.

Tweets were labelled with 4 main classes:
* 0: non-health
* 1: awareness
* 2: other-mention
* 3: self-mention

// add gui images


## App Installation and Usage
```Python
# Install the required packages
pip install requirements.txt
# Run the application
python3 app.py
```

## Model Development
Model was trained on preprocessed tweet data, before finetuning and evaluation against validation and test sets. ERNIE 2.0 was found to be the best performing, reporting an accuracy of 88.5% on the test set. Note that a similarly high performing model based on [DistilBERT](https://huggingface.co/dibsondivya/distilbert-phmtweets-sutd) was also finetuned, performing with 87.7% accuracy. 

## Model Installation and Usage
For ERNIE 2.0 model:
```Python
from transformers import AutoTokenizer, AutoModelForSequenceClassification
tokenizer = AutoTokenizer.from_pretrained("dibsondivya/ernie-phmtweets-sutd")
model = AutoModelForSequenceClassification.from_pretrained("dibsondivya/ernie-phmtweets-sutd")
```

For DistilBERT model:
```Python
from transformers import AutoTokenizer, AutoModelForSequenceClassification
tokenizer = AutoTokenizer.from_pretrained("dibsondivya/distilbert-phmtweets-sutd")
model = AutoModelForSequenceClassification.from_pretrained("dibsondivya/distilbert-phmtweets-sutd")
```

## Dependencies
* Python 3.7 
* transformers
* datasets


## References for Models Attempted
For ERNIE 2.0 Model:
```bibtex
@article{sun2019ernie20,
  title={ERNIE 2.0: A Continual Pre-training Framework for Language Understanding},
  author={Sun, Yu and Wang, Shuohuan and Li, Yukun and Feng, Shikun and Tian, Hao and Wu, Hua and Wang, Haifeng},
  journal={arXiv preprint arXiv:1907.12412},
  year={2019} 
}
```

For DistilBERT model:
```bibtex
@article{Sanh2019DistilBERTAD,
  title={DistilBERT, a distilled version of BERT: smaller, faster, cheaper and lighter},
  author={Victor Sanh and Lysandre Debut and Julien Chaumond and Thomas Wolf},
  journal={ArXiv},
  year={2019},
  volume={abs/1910.01108}
}
```


CHANGELOG.MD
## 12 June 2022
* Linked app to html templates via AJAX and jsonify

bert.py
from torch import nn

class BERT_Arch(nn.Module):
    

    def __init__(self, bert):
      
      super(BERT_Arch, self).__init__()

      self.bert = bert 
      
      # dropout layer
      self.dropout = nn.Dropout(0.2)
      
      # relu activation function
      self.relu =  nn.ReLU()

      # dense layer 1
      self.fc1 = nn.Linear(768,512)
      
      # dense layer 2 (Output layer)
      self.fc2 = nn.Linear(512,4)

      #softmax activation function
      self.softmax = nn.LogSoftmax(dim=1)

    #define the forward pass
    def forward(self, sent_id, mask):

      #pass the inputs to the model  
      _, cls_hs = self.bert(sent_id, attention_mask=mask, return_dict=False)
      
      x = self.fc1(cls_hs)

      x = self.relu(x)

      x = self.dropout(x)

      # output layer
      x = self.fc2(x)
      
      # apply softmax activation
      x = self.softmax(x)

      return x

predictmany.py
import os
# import tensorflow as tf
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import torch
from transformers import AutoModel, BertTokenizerFast

from bert import BERT_Arch

DATASET_PATH = './datasets'
MODEL_PATH = './models'

TRAIN_PATH = os.path.join(DATASET_PATH, 'train_textcleaned.csv')
TEST_PATH = os.path.join(DATASET_PATH, 'test_textcleaned.csv')

device = torch.device('cpu')

## =====================================
## PREPARE THE DATA
## =====================================
df = pd.read_csv(TRAIN_PATH)
df = df.drop(['Unnamed: 0'], axis=1)

# train_df = pd.read_csv(TRAIN_PATH)
# train_df = train_df.drop(['Unnamed: 0'], axis=1)
# train_df.head()

# test_df = pd.read_csv(TEST_PATH)
# test_df = test_df.drop(['Unnamed: 0'], axis=1)
# print(test_df.head())

# x_train = np.array(train_df.tweet)
# y_train = np.array(train_df.label)

# x_test = np.array(test_df.tweet)
# y_test = np.array(test_df.label)

# print(x_test)


## =====================================
## MODEL DEFINITION
## =====================================
# import BERT-base pretrained model
bert = AutoModel.from_pretrained('bert-base-uncased')
model = BERT_Arch(bert)

model = model.to(device)

max_seq_len = 30

train_text, temp_text, train_labels, temp_labels = train_test_split(df['tweet'], df['label'], 
                                                                    random_state=2018, 
                                                                    test_size=0.3, 
                                                                    stratify=df['label'])

# we will use temp_text and temp_labels to create validation and test set
val_text, test_text, val_labels, test_labels = train_test_split(temp_text, temp_labels, 
                                                                random_state=2018, 
                                                                test_size=0.5, 
                                                                stratify=temp_labels)

# Load the BERT tokenizer
tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased')

# tokenize and encode sequences in the test set
tokens_test = tokenizer.batch_encode_plus(
    test_text.tolist(),
    max_length = max_seq_len,
    pad_to_max_length=True,
    truncation=True,
    return_token_type_ids=False
)

test_seq = torch.tensor(tokens_test['input_ids'])
test_mask = torch.tensor(tokens_test['attention_mask'])
test_y = torch.tensor(test_labels.tolist())


## =====================================
## LOAD THE WEIGHTS
## =====================================
# path = './models/saved_weights-500_epochs.pt'
path = './models/model-75-torch.pkl'
model.load_state_dict(torch.load(path, map_location=torch.device('cpu')))

# get predictions for test data
with torch.no_grad():
  preds = model(test_seq.to(device), test_mask.to(device))
  preds = preds.detach().cpu().numpy()
  
# model's performance
preds = np.argmax(preds, axis = 1)
# print(classification_report(test_y, preds))
print(preds)

distilbertpredict.py
from transformers import AutoTokenizer, AutoModelForSequenceClassification

import numpy as np
import torch

device = torch.device('cpu')

def predict(tweet):

    tokenizer = AutoTokenizer.from_pretrained('dibsondivya/distilbert-phm-tweets')
    
    model = AutoModelForSequenceClassification.from_pretrained('dibsondivya/distilbert-phm-tweets')

    # Tokenize inputs
    inputs = tokenizer(tweet, padding=True, truncation=True, return_tensors="pt").to(device) # Move the tensor to the GPU

    # Inference model and get logits
    outputs = model(**inputs)

    # get predictions for test data
    with torch.no_grad():
        predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
        preds = predictions.detach().cpu().numpy()

    print(preds)
    # model's performance
    preds = np.argmax(predictions)
    # print(classification_report(test_y, preds))

    # model's performance
    preds = np.argmax(predictions)
    # print(classification_report(test_y, preds))

    return preds.item()

# pred = predict("Alzheimer's is the worst disease on the planet")
# print(pred)

def convert_label(label):
    label = int(label)

    if label == 0:
        return 'NON-HEALTH'
    elif label == 1:
        return 'AWARENESS'
    elif label == 2:
        return 'OTHER MENTION'
    elif label == 3:
        return 'SELF-MENTION'

predict.py
import os
# import tensorflow as tf
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import torch
from transformers import AutoModel, BertTokenizerFast

from bert import BERT_Arch

DATASET_PATH = './datasets'
MODEL_PATH = './models'

TRAIN_PATH = os.path.join(DATASET_PATH, 'train_textcleaned.csv')
TEST_PATH = os.path.join(DATASET_PATH, 'test_textcleaned.csv')

device = torch.device('cpu')

def predict(tweet):
    ## =====================================
    ## PREPARE THE DATA
    ## =====================================
    df = pd.read_csv(TRAIN_PATH)
    df = df.drop(['Unnamed: 0'], axis=1)

    # train_df = pd.read_csv(TRAIN_PATH)
    # train_df = train_df.drop(['Unnamed: 0'], axis=1)
    # train_df.head()

    # test_df = pd.read_csv(TEST_PATH)
    # test_df = test_df.drop(['Unnamed: 0'], axis=1)
    # print(test_df.head())

    # x_train = np.array(train_df.tweet)
    # y_train = np.array(train_df.label)

    # x_test = np.array(test_df.tweet)
    # y_test = np.array(test_df.label)

    # print(x_test)


    ## =====================================
    ## MODEL DEFINITION
    ## =====================================
    # import BERT-base pretrained model
    bert = AutoModel.from_pretrained('bert-base-uncased')
    model = BERT_Arch(bert)

    model = model.to(device)

    max_seq_len = 30

    train_text, temp_text, train_labels, temp_labels = train_test_split(df['tweet'], df['label'], 
                                                                        random_state=2018, 
                                                                        test_size=0.3, 
                                                                        stratify=df['label'])

    # we will use temp_text and temp_labels to create validation and test set
    val_text, test_text, val_labels, test_labels = train_test_split(temp_text, temp_labels, 
                                                                    random_state=2018, 
                                                                    test_size=0.5, 
                                                                    stratify=temp_labels)

    # Load the BERT tokenizer
    tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased')
    
    # tokenize and encode sequences in the test set
    tokens_test = tokenizer.batch_encode_plus(
        [tweet],
        max_length = max_seq_len,
        pad_to_max_length=True,
        truncation=True,
        return_token_type_ids=False
    )

    test_seq = torch.tensor(tokens_test['input_ids'])
    test_mask = torch.tensor(tokens_test['attention_mask'])
    test_y = torch.tensor(test_labels.tolist())


    ## =====================================
    ## LOAD THE WEIGHTS
    ## =====================================
    # path = './models/saved_weights-500_epochs.pt'
    path = './models/model-75-torch.pkl'
    model.load_state_dict(torch.load(path, map_location=torch.device('cpu')))

    # get predictions for test data
    with torch.no_grad():
        preds = model(test_seq.to(device), test_mask.to(device))
        preds = preds.detach().cpu().numpy()
    
    # model's performance
    preds = np.argmax(preds, axis = 1)
    # print(classification_report(test_y, preds))
    print(preds[0])

    return preds[0].item()

def convert_label(label):
    ## The labels are as follows:
    ## 0: non-health
    ## 1: awareness
    ## 2: other-mention
    ## 3: self-mention
    label = int(label)
    if label == 0:
        return 'Non-health'
    elif label == 1:
        return 'Awareness'
    elif label == 2:
        return 'Other mention'
    elif label == 3:
        return 'Self-mention'